### PR TITLE
build: support passing preprocessor flags to DTS processing

### PR DESCRIPTION
**Pull request by pabigot**
In some cases it's useful to use the C preprocessor to control the content of a devicetree file.  A specific use case is in testing, where different combinations of node properties must be checked (see #27360).  Provide a mechanism to allow arbitrary preprocessor flags to be passed through CMake to affect the devicetree content.

### CODE REVIEW

**Review by mbolivar-nordic — APPROVED** [review]

**Review by henrikbrixandersen — APPROVED** [review]

**Comment by pabigot** [comment]
Thanks for the reviews.  Marking DNM because this needs @galak's input before being merged.

File: cmake/dts.cmake [hunk 1]
```diff
@@ -55,6 +55,11 @@ set(dts_files
   ${shield_dts_files}
   )
 
+# Pass preprocessor flags from the environment
+set(env_dts_cpp $ENV{DTS_EXTRA_CPPFLAGS})
```
Comment (tejlmand) [inline cmake/dts.cmake:59]: This takes settings from environment.
Why is `-DDTS_EXTRA_CPPFLAGS=<flags>` not sufficient ?

In general, Zephyr doesn't allow for passing compiler flags through environment.
https://github.com/zephyrproject-rtos/zephyr/blob/master/cmake/generic_toolchain.cmake#L16-L17

> message(WARNING "The environment variable '${var}' was set to $ENV{${var}},
but Zephyr ignores flags from the environment. Use 'cmake -DEXTRA_${var}=$ENV{${var}}' instead.")

**Review by tejlmand — CHANGES_REQUESTED** [review]

File: cmake/dts.cmake (hunk 1 above)
Comment (pabigot) [reply cmake/dts.cmake:59]: When I searched for how to do this I found examples using this idiom, so I followed it.

There are cases where I'd want to add or set flags in a `CMakeLists.txt` file, as well as externally.  I'll see if I can come up with a way that doesn't involve the environment.

Comment (tejlmand) [reply cmake/dts.cmake:59]: > When I searched for how to do this I found examples using this idiom, so I followed it.

For conf files, I believe, not for preprocessing / compile flags.

**Review by tejlmand — APPROVED** [review]
lgtm

**Review by galak — APPROVED** [review]

**Pull request merged**